### PR TITLE
ENH: Support fold argument in Timestamp.replace

### LINKED ISSUE #25017: Timestamp.replace should handle ambiguous DST times gracefully
```python

>>> tz = pytz.timezone('Europe/London')
>>> d = datetime.datetime(2019, 10, 27, 1, 0)
>>> d = tz.localize(d, is_dst=True)
>>> ts = pd.Timestamp(d)
>>> ts == d 
True
>>> pd.Timestamp(d).replace(minute=0) == d.replace(minute=0)
False
>>> pd.Timestamp(d).replace(minute=0).tzinfo
<DstTzInfo 'Europe/London' GMT0:00:00 STD>
>>> d.replace(minute=0).tzinfo
<DstTzInfo 'Europe/London' BST+1:00:00 DST>

```

This is related to [issue 18319](https://github.com/pandas-dev/pandas/issues/18319) and the associated PR which makes `pandas.Timestamp.replace` behaviour deviate from `datetime.datetime.replace` when crossing DST transitions, and generally hand them more gracefully.
In cases when replacing into an ambiguous time (as above, when the clocks go back and repeat the same time in local time), the default behaviour of pandas is to normalize the datetime, using pytz's default setting `is_dst=False`. It would be useful to 
- have a way of specifying whether to go into DST or not and override the pytz default
- in the case of replacing increments less than an hour, to copy the DST info of the original datetime (since the new time could be assumed to be in the same local hour)

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #25017
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Since `Timestamp` has its own `replace` method, I think we can still introduce this while still supporting PY3.5 (`datetime.replace` gained the `fold` argument in 3.6) while it mimics the functionality in PY3.6

### CODE REVIEW

**Comment by jreback** [comment]
you need to skip a PY2 test I think.

File: pandas/tests/scalar/timestamp/test_unary_ops.py [hunk 1]
```diff
@@ -329,6 +329,18 @@ def test_replace_dst_border(self):
         expected = Timestamp('2013-11-3 03:00:00', tz='America/Chicago')
         assert result == expected
 
+    @pytest.mark.parametrize('fold', [0, 1])
```
Comment (jreback) [inline pandas/tests/scalar/timestamp/test_unary_ops.py]: can you add a test where we pass fold=1 in the constructor itself

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/scalar/timestamp/test_unary_ops.py (hunk 1 above)
Comment (mroeschke) [reply pandas/tests/scalar/timestamp/test_unary_ops.py]: I prefer if we tackle `fold` in the constructor in another PR. `fold` doesn't have any formal handling in the `Timestamp` constructor besides just being carried along by `datetime.datetime`

Comment (mroeschke) [reply pandas/tests/scalar/timestamp/test_unary_ops.py]: https://github.com/pandas-dev/pandas/issues/25057

**Review by jreback — APPROVED** [review]

**Pull request merged**